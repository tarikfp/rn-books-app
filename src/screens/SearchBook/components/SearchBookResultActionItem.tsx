import React from 'react';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import { Box } from '~components/Box';
import { BorderlessBounceButton } from '~components/Button';

type Props = {
  onPress: () => void;
  iconName: string;
  iconColor: string;
  iconSize?: number;
  withBadge?: boolean;
  testID?: string;
};

export const SearchBookResultActionItem: React.FC<Props> = React.memo(
  ({
    iconName,
    iconColor,
    withBadge = true,
    iconSize = 25,
    onPress,
    testID,
  }) => {
    return (
      <BorderlessBounceButton
        testID={testID ?? 'SearchBookResultActionItemButton'}
        activeOpacity={0.65}
        bounceValue={0.85}
        onPress={onPress}
      >
        {withBadge ? (
          <Box borderRadius={'rounded'} padding="xs" backgroundColor={'border'}>
            <MaterialCommunityIcons
              disabled
              name={iconName}
              size={iconSize}
              color={iconColor}
            />
          </Box>
        ) : (
          <MaterialCommunityIcons
            disabled
            name={iconName}
            size={iconSize}
            color={iconColor}
          />
        )}
      </BorderlessBounceButton>
    );
  },
);

SearchBookResultActionItem.displayName = 'SearchBookResultActionItem';
