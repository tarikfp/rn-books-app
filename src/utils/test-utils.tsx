import { ThemeProvider } from '@shopify/restyle';
import { render, RenderOptions } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { theme } from '~theme/core';
import { ThemePreferenceContext } from '~theme/useThemePreference';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemePreferenceContext.Provider
      value={{
        isDarkTheme: false,
        toggleTheme: () => {},
      }}
    >
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemePreferenceContext.Provider>
  );
};

const withAllTheProviders = (ui: React.ReactNode) => (
  <AllTheProviders children={ui} />
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderer = (ui: React.ReactElement) =>
  renderer.create(withAllTheProviders(ui));

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
// override renderer method
export { customRenderer as createRenderer };
