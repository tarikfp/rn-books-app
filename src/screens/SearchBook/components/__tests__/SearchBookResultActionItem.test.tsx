import { fireEvent } from '@testing-library/react-native';
import { createRenderer, render } from '~utils/test-utils';
import { SearchBookResultActionItem } from '../SearchBookResultActionItem';

describe('SearchBookResultActionItem', () => {
  it('renders correctly with badge', () => {
    const tree = createRenderer(
      <SearchBookResultActionItem
        iconName="bookmark"
        iconColor="blue"
        onPress={() => {}}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly without badge', () => {
    const tree = createRenderer(
      <SearchBookResultActionItem
        iconName="bookmark"
        iconColor="blue"
        withBadge={false}
        onPress={() => {}}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('calls onPress when pressed', () => {
    const onPressMock = jest.fn();
    const { getByTestId } = render(
      <SearchBookResultActionItem
        iconName="bookmark"
        iconColor="blue"
        onPress={onPressMock}
      />,
    );

    const button = getByTestId('SearchBookResultActionItemButton');
    fireEvent.press(button);

    expect(onPressMock).toHaveBeenCalled();
  });
});
