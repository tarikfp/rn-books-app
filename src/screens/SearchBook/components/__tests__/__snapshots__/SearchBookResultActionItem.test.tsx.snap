// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBookResultActionItem renders correctly with badge 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.65}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  onPress={[Function]}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
  testID="SearchBookResultActionItemButton"
>
  <View
    style={
      [
        {
          "backgroundColor": "#E7E8E9",
          "borderRadius": 999,
          "padding": 4,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      disabled={true}
      selectable={false}
      style={
        [
          {
            "color": "blue",
            "fontSize": 25,
          },
          undefined,
          {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      󰃀
    </Text>
  </View>
</RNGestureHandlerButton>
`;

exports[`SearchBookResultActionItem renders correctly without badge 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.65}
  animatedStyle={
    {
      "value": {
        "transform": [
          {
            "scale": 1,
          },
        ],
      },
    }
  }
  borderless={true}
  collapsable={false}
  delayLongPress={600}
  handlerTag={2}
  handlerType="NativeViewGestureHandler"
  onActivated={[Function]}
  onActiveStateChange={[Function]}
  onEnded={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  onPress={[Function]}
  style={
    {
      "opacity": 1,
      "transform": [
        {
          "scale": 1,
        },
      ],
    }
  }
  testID="SearchBookResultActionItemButton"
>
  <Text
    allowFontScaling={false}
    disabled={true}
    selectable={false}
    style={
      [
        {
          "color": "blue",
          "fontSize": 25,
        },
        undefined,
        {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    󰃀
  </Text>
</RNGestureHandlerButton>
`;
